src/indexmap.jl

```diff
@@ -68,44 +114,85 @@ function ITensors.inds(imap::IndexMap)
+end
+function real_inds(imap::IndexMap, dimensions::Vector{Int64}=[i for i in 1:dimension(imap)])
   return collect(
-    filter(i -> index_dimension(imap)[i] == dimension, keys(index_dimension(imap)))
+    filter(
+      i -> is_real(imap, i) && dimension(imap, i) ∈ dimensions, keys(index_dimension(imap))
+    ),
   )
 end
-function ind(imap::IndexMap, dimension::Int, digit::Int)
+function imaginary_inds(
+  imap::IndexMap, dimensions::Vector{Int64}=[i for i in 1:dimension(imap)]
+)
+  return collect(
+    filter(
+      i -> !is_real(imap, i) && dimension(imap, i) ∈ dimensions, keys(index_dimension(imap))
+    ),
+  )
+end
+function ind(imap::IndexMap, dim::Int, dig::Int, is_re::Bool=true)
   return only(
     filter(
-      i -> index_dimension(imap)[i] == dimension && index_digit(imap)[i] == digit,
+      i -> dimension(imap, i) == dim && digit(imap, i) == dig && is_real(imap, i) == is_re,
       keys(index_dimension(imap)),
     ),
   )
 end
 
-function dimension_indices(imap::IndexMap, dimap::Int)
-  return filter(ind -> dimension(imap, ind) == dim, inds(imap))
-end
-
 function calculate_xyz(imap::IndexMap, ind_to_ind_value_map, dimensions::Vector{Int})
   out = Number[]
   for dimension in dimensions
-    indices = dimension_inds(imap, dimension)
-    push!(
-      out,
-      sum([index_value_to_scalar(imap, ind, ind_to_ind_value_map[ind]) for ind in indices]),
-    )
+    real_indices, imag_indices = real_inds(imap, [dimension]),
+    imaginary_inds(imap, [dimension])
+    real = sum([
+      index_value_to_scalar(imap, ind, ind_to_ind_value_map[ind]) for ind in real_indices
+    ])
+    imag = if isempty(imag_indices)
+      0.0
+    else
+      sum([
+        index_value_to_scalar(imap, ind, ind_to_ind_value_map[ind]) for ind in imag_indices
+      ])
+    end
+    push!(out, real + imag)
   end
   return out
 end
 
-function calculate_xyz(imap::IndexMap, ind_to_ind_value_map)
-  return calculate_xyz(imap, ind_to_ind_value_map, [i for i in 1:dimension(imap)])
+function calculate_xyz(imap::IndexMap, ind_to_ind_value_map; kwargs...)
+  return calculate_xyz(
+    imap, ind_to_ind_value_map, [i for i in 1:dimension(imap)]; kwargs...
+  )
 end
-function calculate_x(imap::IndexMap, ind_to_ind_value_map, dimension::Int)
-  return only(calculate_xyz(imap, ind_to_ind_value_map, [dimension]))
+function calculate_x(imap::IndexMap, ind_to_ind_value_map, dimension::Int; kwargs...)
+  return only(calculate_xyz(imap, ind_to_ind_value_map, [dimension]; kwargs...))
 end
-function calculate_x(imap::IndexMap, ind_to_ind_value_map)
-  return calculate_x(imap, ind_to_ind_value_map, 1)
+function calculate_x(imap::IndexMap, ind_to_ind_value_map; kwargs...)
+  return calculate_x(imap, ind_to_ind_value_map, 1; kwargs...)
+end
+
+function set_ind_values(
+  imap::IndexMap, sorted_inds::Vector, ind_to_ind_value_map::Dictionary, x::Number
+)
+  ind_to_ind_value_map = copy(ind_to_ind_value_map)
```

Comment:
Seems quite a lot of `copy`s given that every `ind_to_ind_value_map` will do a copy of the dict, maybe this should be a modifying function